File: frontend/src/shared/capsule/codec.ts
/**
 * The 145-bit capsule codec. Pure: no network, no filesystem, no clock (the
 * caller supplies `issueSlot`/`schemaVer`/etc already computed) — CLAUDE.md
 * requires this because the decoder must run forever on a phone in airplane
 * mode, and a corpus of these encode/decode round-trips is meant to run
 * identically in the Python backend's eventual codec/ module and here.
 *
 * `encode` never throws on out-of-range physical input — every quantiser
 * clamps (bits.ts::clampInt). The one thing `encode` DOES throw on is an
 * internal field-table bug, which `fieldTable.ts` already guards at import
 * time.
 */
import { BitReader, BitWriter } from "./bits";
import { boundaryTypeCodec, chlClassCodec, decodeHazardFlags, encodeHazardFlags, msgTypeCodec, quantize, dequantize, riskClassCodec } from "./quantizers";
import { CAPSULE_BYTES, fieldOffsets } from "./fieldTable";
import type { Capsule } from "../types/domain";
import { clampInt } from "./bits";

const OFFSETS = fieldOffsets();
const byName = Object.fromEntries(OFFSETS.map((f) => [f.name, f]));

function widthOf(name: string): number {
  return byName[name].width;
}

export function encodeCapsule(c: Capsule): Uint8Array {
  const w = new BitWriter(CAPSULE_BYTES * 8);

  w.write(msgTypeCodec.encode(c.msgType), widthOf("msg_type"));
  w.write(clampInt(c.schemaVer, 0, 2 ** widthOf("schema_ver") - 1), widthOf("schema_ver"));
  w.write(clampInt(c.issueSlot, 0, 2 ** widthOf("issue_slot") - 1), widthOf("issue_slot"));
  w.write(clampInt(c.validHours, 0, 2 ** widthOf("valid_hours") - 1), widthOf("valid_hours"));
  w.write(clampInt(c.zoneId, 0, 2 ** widthOf("zone_id") - 1), widthOf("zone_id"));
  w.write(clampInt(c.latOffset, 0, 255), widthOf("lat_offset"));
  w.write(clampInt(c.lonOffset, 0, 255), widthOf("lon_offset"));
  w.write(riskClassCodec.encode(c.riskClass), widthOf("risk_class"));
  w.write(encodeHazardFlags(c.hazardFlags), widthOf("hazard_flags"));
  w.write(quantize.waveHs(c.waveHs), widthOf("wave_hs"));
  w.write(quantize.windKt(c.windKt), widthOf("wind_kt"));
  w.write(clampInt(c.windDir16, 0, 15), widthOf("wind_dir"));
  w.write(quantize.currKt(c.currKt), widthOf("curr_kt"));
  w.write(clampInt(c.currDir16, 0, 15), widthOf("curr_dir"));
  w.write(quantize.sst(c.sstC), widthOf("sst"));
  w.write(chlClassCodec.encode(c.chlClass), widthOf("chl_class"));
  w.write(clampInt(c.pfzBearing16, 0, 15), widthOf("pfz_bearing"));
  w.write(quantize.pfzDistNm(c.pfzDistNm), widthOf("pfz_dist_nm"));
  w.write(clampInt(c.pfzConfidence, 0, 3), widthOf("pfz_conf"));
  w.write(quantize.bndDistNm(c.bndDistNm), widthOf("bnd_dist_nm"));
  w.write(boundaryTypeCodec.encode(c.bndType), widthOf("bnd_type"));
  w.write(quantize.bndEtaMin(c.bndEtaMin), widthOf("bnd_eta_min"));
  w.write(clampInt(c.reasonCode, 0, 255), widthOf("reason_code"));
  w.write(parseEvidenceHash(c.evidenceHash), widthOf("evidence_hash"));

  return w.toBytes();
}

export function decodeCapsule(bytes: Uint8Array): Capsule {
  if (bytes.length < CAPSULE_BYTES) {
    throw new Error(`capsule must be at least ${CAPSULE_BYTES} bytes, got ${bytes.length}`);
  }
  const r = new BitReader(bytes);

  const msgType = msgTypeCodec.decode(r.read(widthOf("msg_type")));
  const schemaVer = r.read(widthOf("schema_ver"));
  const issueSlot = r.read(widthOf("issue_slot"));
  const validHours = r.read(widthOf("valid_hours"));
  const zoneId = r.read(widthOf("zone_id"));
  const latOffset = r.read(widthOf("lat_offset"));
  const lonOffset = r.read(widthOf("lon_offset"));
  const riskClass = riskClassCodec.decode(r.read(widthOf("risk_class")));
  const hazardFlags = decodeHazardFlags(r.read(widthOf("hazard_flags")));
  const waveHs = dequantize.waveHs(r.read(widthOf("wave_hs")));
  const windKt = dequantize.windKt(r.read(widthOf("wind_kt")));
  const windDir16 = r.read(widthOf("wind_dir"));
  const currKt = dequantize.currKt(r.read(widthOf("curr_kt")));
  const currDir16 = r.read(widthOf("curr_dir"));
  const sstC = dequantize.sst(r.read(widthOf("sst")));
  const chlClass = chlClassCodec.decode(r.read(widthOf("chl_class")));
  const pfzBearing16 = r.read(widthOf("pfz_bearing"));
  const pfzDistNm = dequantize.pfzDistNm(r.read(widthOf("pfz_dist_nm")));
  const pfzConfidence = r.read(widthOf("pfz_conf")) as 0 | 1 | 2 | 3;
  const bndDistNm = dequantize.bndDistNm(r.read(widthOf("bnd_dist_nm")));
  const bndType = boundaryTypeCodec.decode(r.read(widthOf("bnd_type")));
  const bndEtaMin = dequantize.bndEtaMin(r.read(widthOf("bnd_eta_min")));
  const reasonCode = r.read(widthOf("reason_code"));
  const evidenceHash = formatEvidenceHash(r.read(widthOf("evidence_hash")));

  return {
    msgType, schemaVer, issueSlot, validHours, zoneId, latOffset, lonOffset,
    riskClass, hazardFlags, waveHs, windKt, windDir16, currKt, currDir16,
    sstC, chlClass, pfzBearing16, pfzDistNm, pfzConfidence, bndDistNm,
    bndType, bndEtaMin, reasonCode, evidenceHash,
  };
}

/** Accepts a full SHA-256 hex digest or an already-truncated 3-hex-char
 * fingerprint; only the top 12 bits ever reach the wire. */
function parseEvidenceHash(hash: string): number {
  const prefix = hash.slice(0, 3).padEnd(3, "0");
  return clampInt(parseInt(prefix, 16), 0, 4095);
}

function formatEvidenceHash(raw: number): string {
  return raw.toString(16).padStart(3, "0");
}

/** SHA-256 over a canonical JSON encoding of the evidence bundle, truncated to
 * the 3 hex chars (12 bits) the capsule carries. Async because Web Crypto is;
 * callers build the Capsule after awaiting this once per advisory. */
export async function computeEvidenceFingerprint(evidence: unknown): Promise<string> {
  const canonical = JSON.stringify(evidence, Object.keys(evidence as object).sort());
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(canonical));
  const hex = Array.from(new Uint8Array(digest)).map((b) => b.toString(16).padStart(2, "0")).join("");
  return hex.slice(0, 3);
}

File: backend/dhruva/codec/codec.py
"""The 145-bit capsule codec — Python half of the pair with
frontend/src/shared/capsule/codec.ts. `tests/test_codec_corpus.py` runs the
same `eval/capsule_corpus.json` fixtures the TypeScript test suite runs,
proving the two implementations agree byte-for-byte, per CLAUDE.md: "Python
and TypeScript run the same corpus in CI."
"""

from __future__ import annotations

from dataclasses import dataclass, field

from dhruva.codec.bits import BitReader, BitWriter, clamp_int
from dhruva.codec.enums import (
    BOUNDARY_TYPE_VALUES,
    CHL_CLASS_VALUES,
    MSG_TYPE_VALUES,
    RISK_CLASS_VALUES,
    decode_enum,
    decode_hazard_flags,
    encode_enum,
    encode_hazard_flags,
)
from dhruva.codec.field_table import CAPSULE_BYTES, CAPSULE_TOTAL_BITS, WIDTH_BY_NAME
from dhruva.codec.quantizers import (
    dequantize_bnd_dist_nm,
    dequantize_bnd_eta_min,
    dequantize_curr_kt,
    dequantize_pfz_dist_nm,
    dequantize_sst,
    dequantize_wave_hs,
    dequantize_wind_kt,
    quantize_bnd_dist_nm,
    quantize_bnd_eta_min,
    quantize_curr_kt,
    quantize_pfz_dist_nm,
    quantize_sst,
    quantize_wave_hs,
    quantize_wind_kt,
)

__all__ = ["Capsule", "encode_capsule", "decode_capsule"]


@dataclass(slots=True)
class Capsule:
    msg_type: str = "advisory"
    schema_ver: int = 1
    issue_slot: int = 0
    valid_hours: int = 6
    zone_id: int = 0
    lat_offset: int = 0
    lon_offset: int = 0
    risk_class: str = "unknown"
    hazard_flags: list[str] = field(default_factory=list)
    wave_hs: float = 0.0
    wind_kt: float = 0.0
    wind_dir16: int = 0
    curr_kt: float = 0.0
    curr_dir16: int = 0
    sst_c: float = 18.0
    chl_class: str = "low"
    pfz_bearing16: int = 0
    pfz_dist_nm: float = 0.0
    pfz_confidence: int = 0
    bnd_dist_nm: float = 0.0
    bnd_type: str = "none"
    bnd_eta_min: float = 0.0
    reason_code: int = 0
    evidence_hash: str = "000"


def _w(name: str) -> int:
    return WIDTH_BY_NAME[name]


def encode_capsule(c: Capsule) -> bytes:
    w = BitWriter(CAPSULE_BYTES * 8)
    w.write(encode_enum(MSG_TYPE_VALUES, c.msg_type), _w("msg_type"))
    w.write(clamp_int(c.schema_ver, 0, 2 ** _w("schema_ver") - 1), _w("schema_ver"))
    w.write(clamp_int(c.issue_slot, 0, 2 ** _w("issue_slot") - 1), _w("issue_slot"))
    w.write(clamp_int(c.valid_hours, 0, 2 ** _w("valid_hours") - 1), _w("valid_hours"))
    w.write(clamp_int(c.zone_id, 0, 2 ** _w("zone_id") - 1), _w("zone_id"))
    w.write(clamp_int(c.lat_offset, 0, 255), _w("lat_offset"))
    w.write(clamp_int(c.lon_offset, 0, 255), _w("lon_offset"))
    w.write(encode_enum(RISK_CLASS_VALUES, c.risk_class), _w("risk_class"))
    w.write(encode_hazard_flags(c.hazard_flags), _w("hazard_flags"))
    w.write(quantize_wave_hs(c.wave_hs), _w("wave_hs"))
    w.write(quantize_wind_kt(c.wind_kt), _w("wind_kt"))
    w.write(clamp_int(c.wind_dir16, 0, 15), _w("wind_dir"))
    w.write(quantize_curr_kt(c.curr_kt), _w("curr_kt"))
    w.write(clamp_int(c.curr_dir16, 0, 15), _w("curr_dir"))
    w.write(quantize_sst(c.sst_c), _w("sst"))
    w.write(encode_enum(CHL_CLASS_VALUES, c.chl_class), _w("chl_class"))
    w.write(clamp_int(c.pfz_bearing16, 0, 15), _w("pfz_bearing"))
    w.write(quantize_pfz_dist_nm(c.pfz_dist_nm), _w("pfz_dist_nm"))
    w.write(clamp_int(c.pfz_confidence, 0, 3), _w("pfz_conf"))
    w.write(quantize_bnd_dist_nm(c.bnd_dist_nm), _w("bnd_dist_nm"))
    w.write(encode_enum(BOUNDARY_TYPE_VALUES, c.bnd_type), _w("bnd_type"))
    w.write(quantize_bnd_eta_min(c.bnd_eta_min), _w("bnd_eta_min"))
    w.write(clamp_int(c.reason_code, 0, 255), _w("reason_code"))
    w.write(_parse_evidence_hash(c.evidence_hash), _w("evidence_hash"))
    return w.to_bytes()


def decode_capsule(data: bytes) -> Capsule:
    if len(data) < CAPSULE_BYTES:
        raise ValueError(f"capsule must be at least {CAPSULE_BYTES} bytes, got {len(data)}")
    r = BitReader(data)
    c = Capsule()
    c.msg_type = decode_enum(MSG_TYPE_VALUES, r.read(_w("msg_type")))
    c.schema_ver = r.read(_w("schema_ver"))
    c.issue_slot = r.read(_w("issue_slot"))
    c.valid_hours = r.read(_w("valid_hours"))
    c.zone_id = r.read(_w("zone_id"))
    c.lat_offset = r.read(_w("lat_offset"))
    c.lon_offset = r.read(_w("lon_offset"))
    c.risk_class = decode_enum(RISK_CLASS_VALUES, r.read(_w("risk_class")))
    c.hazard_flags = decode_hazard_flags(r.read(_w("hazard_flags")))
    c.wave_hs = dequantize_wave_hs(r.read(_w("wave_hs")))
    c.wind_kt = dequantize_wind_kt(r.read(_w("wind_kt")))
    c.wind_dir16 = r.read(_w("wind_dir"))
    c.curr_kt = dequantize_curr_kt(r.read(_w("curr_kt")))
    c.curr_dir16 = r.read(_w("curr_dir"))
    c.sst_c = dequantize_sst(r.read(_w("sst")))
    c.chl_class = decode_enum(CHL_CLASS_VALUES, r.read(_w("chl_class")))
    c.pfz_bearing16 = r.read(_w("pfz_bearing"))
    c.pfz_dist_nm = dequantize_pfz_dist_nm(r.read(_w("pfz_dist_nm")))
    c.pfz_confidence = r.read(_w("pfz_conf"))
    c.bnd_dist_nm = dequantize_bnd_dist_nm(r.read(_w("bnd_dist_nm")))
    c.bnd_type = decode_enum(BOUNDARY_TYPE_VALUES, r.read(_w("bnd_type")))
    c.bnd_eta_min = dequantize_bnd_eta_min(r.read(_w("bnd_eta_min")))
    c.reason_code = r.read(_w("reason_code"))
    c.evidence_hash = _format_evidence_hash(r.read(_w("evidence_hash")))
    return c


def _parse_evidence_hash(hash_str: str) -> int:
    prefix = (hash_str[:3] or "0").ljust(3, "0")
    return clamp_int(int(prefix, 16), 0, 4095)


def _format_evidence_hash(raw: int) -> str:
    return format(raw, "x").rjust(3, "0")


assert CAPSULE_TOTAL_BITS == 145  # documents the invariant at import time too
